src/compare-pdf-to-snapshot.ts

```
@@ -137,61 +135,111 @@ export const snapshotsDirName = '__snapshots__'
-        if (result.equal) {
-          if (existsSync(diffSnapshotPath)) {
-            unlinkSync(diffSnapshotPath)
-          }
-
-          if (existsSync(newSnapshotPath)) {
-            unlinkSync(newSnapshotPath)
-          }
-
-          return true
-        }
+  const basePath = path.join(dirPath, snapshotName)
+
+  return {
+    name: snapshotName,
+    dirPath,
+    path: `${basePath}.png`,
+    diffPath: `${basePath}.diff.png`,
+    newPath: `${basePath}.new.png`,
+  }
+}
 
-        return writeImages(newSnapshotPath)(images)
-          .then(() => writeImages(diffSnapshotPath)(result.diffs.map((x) => x.diff)))
-          .then(() => false)
-      }),
-    )
+async function handleMissingSnapshot(
+  pdf: string | Buffer,
+  snapshotContext: SnapshotContext,
+  { failOnMissingSnapshot, maskRegions, pdf2PngOptions }: Required<CompareOptions>,
+): Promise<boolean> {
+  if (failOnMissingSnapshot) {
+    return false
+  }
+
+  // Generate snapshot if missing
+  const images = await pdf2png(pdf, pdf2PngOptions).then(maskImgWithRegions(maskRegions))
+  await writeImages(snapshotContext.path)(images)
+
+  return true
+}
+
+async function compareWithSnapshot(
+  pdf: string | Buffer,
+  snapshotContext: SnapshotContext,
+  { maskRegions, pdf2PngOptions, tolerance }: Required<CompareOptions>,
+): Promise<boolean> {
+  const images = await pdf2png(pdf, pdf2PngOptions).then(maskImgWithRegions(maskRegions))
+  const result = await compareImages(snapshotContext.path, images, { tolerance })
+
+  if (result.equal) {
+    await removeIfExists(snapshotContext.diffPath)
+    await removeIfExists(snapshotContext.newPath)
+
+    return true
+  }
+
+  await writeImages(snapshotContext.newPath)(images)
+  await writeImages(snapshotContext.diffPath)(result.diffs.map((x) => x.diff))
+
+  return false
+}
+
+async function removeIfExists(filePath: string): Promise<void> {
+  try {
+    await unlink(filePath)
+  } catch {
+    // File doesn't exist, no need to remove
```

minor thing but this would technically swallow potentially useful errors, such as `EACCES` or `EBUSY` - errors surfaced from underlying [Linux unlink](https://man7.org/linux/man-pages/man2/unlink.2.html#top_of_page).

May be worth doing a simple [existsSync](https://nodejs.org/docs/latest-v18.x/api/fs.html#fsexistssyncpath) check and conditionally executing unlink based on that?

Throwing hard on the `unlink` op call makes a bit more sense to my brain:

```ts
// async version of this deprecated but sync version still LTS
const fileExists = fs.existsSync(filePath);
if (!fileExists) {
  return;
}
await unlink(filePath);
```

Can still use try and catch if you want more fine tuned control i.e. to ignore specific errors